internal/graph/memory.go

```diff
@@ -101,34 +87,58 @@ func (m *Memory) IndexAll(ctx context.Context, fetcher content.Fetcher, node oci
 // it does not necessarily correspond to any consistent snapshot of the stored
 // contents.
 func (m *Memory) Predecessors(_ context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
 	key := descriptor.FromOCI(node)
-	value, exists := m.predecessors.Load(key)
+	set, exists := m.predecessors[key]
 	if !exists {
 		return nil, nil
 	}
-	predecessors := value.(*sync.Map)
-
 	var res []ocispec.Descriptor
-	predecessors.Range(func(key, value interface{}) bool {
-		res = append(res, value.(ocispec.Descriptor))
-		return true
-	})
+	for k := range set {
+		res = append(res, m.nodes[k])
+	}
 	return res, nil
 }
 
+// Remove removes the node from its predecessors and successors.
+func (m *Memory) Remove(ctx context.Context, node ocispec.Descriptor) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	nodeKey := descriptor.FromOCI(node)
+	// remove the node from its successors' predecessor list
+	for successorKey := range m.successors[nodeKey] {
+		m.predecessors[successorKey].Delete(nodeKey)
+	}
+	delete(m.successors, nodeKey)
```

Comment:
Should we also document that we don't remove the entry of `m.predecessors[nodeKey]`?

Reply:
Refined the `Remove` logic (delete the predecessor entry only if no predecessors of the node still exists. If any predecessors of the node is still predsent, the entry is present) and elaborately documented the behavior in the comment.